File: web/src/api/department.ts
/**
 * 组织架构管理 API
 * 
 * 需求：
 * - 获取部门树
 * - 查看部门下的用户
 * - 给用户分配组织架构（部门和 Leader）
 * - 分配 Leader
 * 
 * 设计思路：
 * - 使用树形结构展示部门层级
 * - 支持点击部门查看该部门下的用户
 * - 支持批量分配用户到部门
 * - 支持为用户分配 Leader
 */

import { get, post, put, del } from '@/utils/request'
import type { UserInfo } from '@/types'

// ==================== 类型定义 ====================

export interface Department {
  id: number
  name: string
  code: string
  parent_id: number | null // NULL 表示根部门
  full_code_path: string
  full_name_path?: string // 完整名称路径（如：技术部/后端组）
  managers: string // 部门负责人（多个用户名逗号分隔）
  description: string
  status: 'active' | 'inactive'
  sort: number
  is_system_default?: boolean // 是否为系统默认组织（不可删除）
  created_at: string
  updated_at: string
  children?: Department[] // 子部门
}

export interface CreateDepartmentReq {
  name: string
  code: string
  parent_id: number
  description?: string
}

export interface CreateDepartmentResp {
  department: Department
}

export interface UpdateDepartmentReq {
  name?: string
  description?: string
  managers?: string // 多个用户名逗号分隔
  status?: 'active' | 'inactive'
  sort?: number
}

export interface UpdateDepartmentResp {
  department: Department
}

export interface GetDepartmentTreeResp {
  departments: Department[]
}

export interface GetDepartmentResp {
  department: Department
}

// ==================== 用户分配相关 ====================

export interface AssignUserReq {
  username: string
  department_full_path?: string // 部门完整路径（可选，为空表示移除部门）
  leader_username?: string // Leader 用户名（可选，为空表示移除 Leader）
}

export interface AssignUserResp {
  user: UserInfo
}

// ==================== API 调用 ====================

/**
 * 获取部门树
 */
export function getDepartmentTree() {
  return get<GetDepartmentTreeResp>('/hr/api/v1/department/tree')
}

/**
 * 根据ID获取部门详情
 */
export function getDepartmentById(id: number) {
  return get<GetDepartmentResp>(`/hr/api/v1/department/${id}`)
}

/**
 * 根据完整路径获取部门详情
 */
export interface GetDepartmentByPathReq {
  full_code_path: string
}

export function getDepartmentByPath(fullCodePath: string) {
  // 先获取部门树，然后查找对应的部门
  return getDepartmentTree().then(res => {
    const findDepartment = (depts: Department[], path: string): Department | null => {
      for (const dept of depts) {
        if (dept.full_code_path === path) {
          return dept
        }
        if (dept.children) {
          const found = findDepartment(dept.children, path)
          if (found) return found
        }
      }
      return null
    }
    
    const department = findDepartment(res.departments, fullCodePath)
    if (!department) {
      throw new Error(`部门不存在: ${fullCodePath}`)
    }
    return { department }
  })
}

/**
 * 创建部门
 */
export function createDepartment(data: CreateDepartmentReq) {
  return post<CreateDepartmentResp>('/hr/api/v1/department', data)
}

/**
 * 更新部门
 */
export function updateDepartment(id: number, data: UpdateDepartmentReq) {
  return put<UpdateDepartmentResp>(`/hr/api/v1/department/${id}`, data)
}

/**
 * 删除部门
 */
export function deleteDepartment(id: number) {
  return del(`/hr/api/v1/department/${id}`)
}

// ==================== 用户分配 API ====================

/**
 * 分配用户组织架构（部门和 Leader）
 */
export function assignUser(data: AssignUserReq) {
  return post<AssignUserResp>('/hr/api/v1/user/assign', data)
}


/**
 * 根据部门完整路径获取用户列表
 */
export interface GetUsersByDepartmentResp {
  users: UserInfo[]
}

export function getUsersByDepartment(departmentFullPath: string) {
  return get<GetUsersByDepartmentResp>('/hr/api/v1/user/department', {
    department_full_path: departmentFullPath
  })
}

File: dto/department.go
package dto

import "github.com/ai-agent-os/ai-agent-os/core/hr-server/model"

// CreateDepartmentReq 创建部门请求
type CreateDepartmentReq struct {
	Name        string `json:"name" binding:"required" example:"技术部"`        // 部门名称
	Code        string `json:"code" binding:"required" example:"tech"`        // 部门编码
	ParentID    int64  `json:"parent_id" example:"0"`                         // 父部门ID（0表示根部门）
	Description string `json:"description" example:"负责技术研发"`                // 部门描述
	Managers    string `json:"managers,omitempty" example:"zhangsan,lisi"`   // 部门负责人（可选，多个用户名逗号分隔）
}

// CreateDepartmentResp 创建部门响应
type CreateDepartmentResp struct {
	Department *model.Department `json:"department"` // 部门信息
}

// UpdateDepartmentReq 更新部门请求
type UpdateDepartmentReq struct {
	Name        *string `json:"name,omitempty" example:"技术部"`              // 部门名称（可选）
	Description *string `json:"description,omitempty" example:"负责技术研发"`  // 部门描述（可选）
	Managers    *string `json:"managers,omitempty" example:"zhangsan,lisi"`  // 部门负责人（可选，多个用户名逗号分隔）
	Status      *string `json:"status,omitempty" example:"active"`          // 状态（可选）：active(激活), inactive(停用)
	Sort        *int    `json:"sort,omitempty" example:"0"`                 // 排序（可选）
}

// UpdateDepartmentResp 更新部门响应
type UpdateDepartmentResp struct {
	Department *model.Department `json:"department"` // 部门信息
}

// GetDepartmentTreeResp 获取部门树响应
type GetDepartmentTreeResp struct {
	Departments []*model.Department `json:"departments"` // 部门树
}

// GetDepartmentResp 获取部门响应
type GetDepartmentResp struct {
	Department *model.Department `json:"department"` // 部门信息
}

// AssignUserReq 分配用户组织架构请求
type AssignUserReq struct {
	Username           string  `json:"username" binding:"required" example:"zhangsan"`           // 用户名
	DepartmentFullPath *string `json:"department_full_path,omitempty" example:"/tech/backend"` // 部门完整路径（可选，为空表示移除部门）
	LeaderUsername     *string `json:"leader_username,omitempty" example:"lisi"`                // Leader 用户名（可选，为空表示移除 Leader）
}

// AssignUserResp 分配用户组织架构响应
type AssignUserResp struct {
	User UserInfo `json:"user"` // 用户信息
}

// GetUsersByDepartmentReq 根据部门获取用户请求
type GetUsersByDepartmentReq struct {
	DepartmentFullPath string `json:"department_full_path" form:"department_full_path" binding:"required" example:"/tech/backend"` // 部门完整路径
}

// GetUsersByDepartmentResp 根据部门获取用户响应
type GetUsersByDepartmentResp struct {
	Users []UserInfo `json:"users"` // 用户列表
}

